registry/storage/tagstore.go

```diff
@@ -132,46 +148,117 @@ func (ts *tagStore) linkedBlobStore(ctx context.Context, tag string) *linkedBlob
 	}
 }
 
+type atomicError struct {
+	mu  sync.Mutex
+	err error
+}
+
+func (e *atomicError) Store(err error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	e.err = err
+}
+
+func (e *atomicError) Load() error {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	return e.err
+}
+
 // Lookup recovers a list of tags which refer to this digest.  When a manifest is deleted by
 // digest, tag entries which point to it need to be recovered to avoid dangling tags.
-func (ts *tagStore) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+func (ts *TagStore) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
 	allTags, err := ts.All(ctx)
 	switch err.(type) {
+	case nil:
+		break
 	case distribution.ErrRepositoryUnknown:
 		// This tag store has been initialized but not yet populated
 		break
-	case nil:
-		break
 	default:
 		return nil, err
 	}
 
-	var tags []string
-	for _, tag := range allTags {
-		tagLinkPathSpec := manifestTagCurrentPathSpec{
-			name: ts.repository.Named().Name(),
-			tag:  tag,
-		}
+	lookupErr := &atomicError{}
+
+	inputChan := make(chan string)
+	outputChan := make(chan string, len(allTags))
 
-		tagLinkPath, _ := pathFor(tagLinkPathSpec)
-		tagDigest, err := ts.blobStore.readlink(ctx, tagLinkPath)
-		if err != nil {
-			switch err.(type) {
-			case storagedriver.PathNotFoundError:
-				continue
+	workersWaitGroup := sync.WaitGroup{}
+	workersWaitGroup.Add(ts.lookupConcurrencyFactor)
+
+	for i := 0; i < ts.lookupConcurrencyFactor; i++ {
```

Comment:
I have a small issue with this. I dont think we need to fire up all `lookupConcurrencyFactor` goroutines at once. Say we only have 2 digests to look up for a given tag - this would be an unnecessary runtime "waste". Ideally we implement this in a sort of goroutine rate limiter fashion

Reply:
Rate limiter approach was the base of this: https://github.com/distribution/distribution/pull/3589
Say we have thousands of tags to lookup isn't it better to spawn a worker pool of a fixed size for this task instead of spawning a new goroutine each time. I don't know if there a good hybrid approach exists.